src/core/builder.rs

```diff
@@ -31,31 +34,94 @@ pub struct GraphBuilder {
     processing: HashSet<String>,
     verbose: bool,
     cache: Option<DepGraph>,
-    metadata: BuildMetadata
+    metadata: BuildMetadata,
+    // For profiling
+    pyro_server: Option<String>,
+    pyro_app_name: String,
+    pyro_invocation_count: u32,
 }
 
 #[pymethods]
 impl GraphBuilder {
     #[new]
-    pub fn new(verbose: Option<bool>) -> Self {
+    pub fn new(
+        verbose: Option<bool>,
+        pyro_server: Option<String>,
+        pyro_app_name: Option<String>,
+    ) -> Self {
         let verbose = if let Some(verbose) = verbose {
             verbose
         } else {
             false
         };
 
+        let pyro_app_name = if let Some(pyro_app_name) = pyro_app_name {
+            pyro_app_name
+        } else {
+            "fast-dep".to_string()
+        };
+
         let builder = GraphBuilder {
             graph: DepGraph::new(),
             processing: HashSet::new(),
             verbose: verbose,
             cache: None,
             metadata: BuildMetadata::new(),
+            // For profiling
+            pyro_server: pyro_server,
+            pyro_app_name: pyro_app_name,
+            pyro_invocation_count: 0,
         };
 
         builder
     }
 
     pub fn build(&mut self, source: &str, package: Option<String>) -> DepGraph {
+        let profiling_agent = if let Some(pyro_server) = self.pyro_server.clone() {
+            let session_name = format!("{}-{}", self.pyro_app_name, 
+            self.pyro_invocation_count);
+            info!(
+                "Building pyroscope agent: {}",
+                session_name
+            );
+            let agent = PyroscopeAgent::builder(
+                pyro_server,
+                session_name.clone()
+            ).backend(pprof_backend(PprofConfig::new().sample_rate(100)))
+            .build();
+
+            let agent = match agent {
+                Ok(agent) => Some(agent),
+                Err(error) => {
+                    warn!("Pyroscope agent build failed with error: {}", error);
+                    None
```

Comment:
Trying to less "frowned on" by using unwraps here but this branching is odd. Ideally here I would just immediately set the outer `profiling_agent` to None, but I can't figure out how to do this. Instead, I have to react to this `None` value in code below, even though I know what I want `profiling_agent` to be and don't need to continue this if clause at all.